### PR TITLE
[WIP] Firewall rules hardening

### PR DESCRIPTION
**Pull request by paramite**
- adds source information to firewall rules

### CODE REVIEW

File: lib/filters/split.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        raise AnsibleFilterError(
+            "Failed to split {} with {} due to:\n{}".format(data, separator, ex)
+        )
+
+
+class FilterModule(object):
+    def filters(self):
+        return {
+            'split': split
+        }
```
Comment (larsks) [inline lib/filters/split.py]: I'm not sure I see why we need to introduce a `split` filter?  You can just call the `split` method on strings, e.g.:

```
{% for thing in list_of_things.split(',') %}
```

File: roles/elasticsearch/tasks/main.yml [hunk 2]
```diff
@@ -39,4 +39,9 @@
 - name: register elasticsearch firewall ports
   set_fact:
     firewall_ports: >
-      {{ firewall_ports + [elasticsearch_port] }}
+      {{ firewall_ports }} + [
+        "{{ item }}:{{ elasticsearch_port }}"
+      ]
+  with_items:
+    - "{{ groups['kibana_hosts']|host_addresses(hostvars, 'ipv4') }}"
```
Comment (larsks) [inline roles/elasticsearch/tasks/main.yml]: I'm having second thoughts about including this sort of logic in the roles themselves.  I am thinking that the roles should be usable in isolation, and logic like this should go into `group_vars` somewhere (e.g., `group_vars/elastic_hosts.yml` or something).

Separately, there should be a configuration parameter for the role that is "the ip address to which clients should connect", and this is the address that should be used in the firewall -- not a list of default addresses from the hosts.

File: roles/firewall/commit/tasks/main.yml [hunk 3]
```diff
@@ -3,15 +3,17 @@
   iptables:
       chain: INPUT
       protocol: tcp
-      destination_port: "{{ item }}"
+      destination_port: "{{ (item|split(':')).1 }}"
```
Comment (larsks) [inline roles/firewall/commit/tasks/main.yml]: No filter necessary:

```
destination_port: "{{ item.split(':').1 }}"
```

File: roles/grafana/tasks/grafana.yml [hunk 4]
```diff
@@ -33,4 +33,8 @@
 - name: insert firewalld rule for grafana
   set_fact:
     firewall_ports: >
-      {{ firewall_ports + [grafana_port] }}
+      {{ firewall_ports }} + [
+        "{{ item }}:{{ grafana_port }}"
+      ]
+  with_items:
+    - '0.0.0.0'
```
Comment (larsks) [inline roles/grafana/tasks/grafana.yml]: Should we be hard coding this here?

File: roles/opstoolsvhost/tasks/main.yml [hunk 5]
```diff
@@ -15,7 +15,9 @@
...
     firewall_ports: >
-      {{ firewall_ports + [item] }}
+      {{ firewall_ports }} + [
+        "{{ item }}:{{ opstools_apache_http_port }}",
+        "{{ item }}:{{ opstools_apache_https_port }}",
+      ]
   with_items:
-    - '{{ opstools_apache_http_port }}'
-    - '{{ opstools_apache_https_port }}'
+    - '0.0.0.0'
```
Comment (larsks) [inline roles/opstoolsvhost/tasks/main.yml]: Ditto.

**Review by larsks — CHANGES_REQUESTED** [review]

File: lib/filters/split.py (hunk 1 above)
Comment (paramite) [reply lib/filters/split.py]: Oh really, last time I tried that, it was not working. Will remove it then.

File: roles/elasticsearch/tasks/main.yml (hunk 2 above)
Comment (paramite) [reply roles/elasticsearch/tasks/main.yml]: That's a fair point. I did not think much about deployment on multi-port hosts and this automation does not make sense in that case. OK, will refactor.

File: roles/grafana/tasks/grafana.yml (hunk 4 above)
Comment (paramite) [reply roles/grafana/tasks/grafana.yml]: :) ... now that you are asking, I don't think that we should.

File: .gitignore [hunk 6]
```diff
@@ -0,0 +1,4 @@
+*.py[co]
+.eggs/
+.tox/
+*.egg-info/
```
Comment (larsks) [inline .gitignore:4]: This all seems fine, but maybe just make it a separate PR and we can just approve it right away?

Comment (larsks) [inline .gitignore:4]: This looks fine...maybe just put it in a separate PR, and we can just merge it immediately?

File: roles/firewall/commit/tasks/main.yml [hunk 7]
```diff
@@ -3,15 +3,17 @@
   iptables:
       chain: INPUT
       protocol: tcp
-      destination_port: "{{ item }}"
+      destination_port: "{{ item.split(':').1 }}"
+      source: "{{ item.split(':').0 }}"
```
Comment (larsks) [inline roles/firewall/commit/tasks/main.yml]: Shouldn't this be "destination", rather than "source"?

File: roles/firewall/commit/tasks/main.yml [hunk 8]
```diff
@@ -3,15 +3,17 @@
...
       jump: ACCEPT
   with_items: '{{ firewall_ports }}'
   when: firewall_manage_rules and firewall_use_iptables
   notify: persist iptables
 
 - name: Enable service ports via firewalld
   firewalld:
-      port: "{{ item }}/tcp"
+      port: "{{ item.split(':').1 }}/tcp"
+      source: "{{ item.split(':').0 }}"
```
Comment (larsks) [inline roles/firewall/commit/tasks/main.yml]: ...also here...

File: roles/grafana/tasks/grafana.yml [hunk 9]
```diff
@@ -33,4 +35,8 @@
 - name: insert firewalld rule for grafana
   set_fact:
     firewall_ports: >
-      {{ firewall_ports + [grafana_port] }}
+      {{ firewall_ports }} + [
+        "{{ item }}:{{ grafana_port }}"
+      ]
+  with_items:
+    - "{{ grafana_bind_address }}"
```
Comment (larsks) [inline roles/grafana/tasks/grafana.yml:46]: We shouldn't be using the bind address for firewall rules.  We should be using whichever address clients will be connecting to...

File: roles/grafana/tasks/graphite.yml [hunk 10]
```diff
@@ -1,17 +1,17 @@
...
-      name: '{{ item }}'
-      state: present
+    name: '{{ item }}'
+    state: present
   with_items:
-      - graphite-web
-      - httpd
-      - python-carbon
+    - graphite-web
+    - httpd
```
Comment (larsks) [inline roles/grafana/tasks/graphite.yml]: We shouldn't be installing httpd here.  There is an `httpd` role for that.

File: roles/grafana/tasks/graphite.yml [hunk 11]
```diff
@@ -20,34 +20,38 @@
...
-      dest: '{{ httpd_config_parts_dir }}/15-graphite-web.conf'
-      create: yes
-      state: present
-      insertafter: "# Graphite Web Basic mod_wsgi vhost"
-      line: "Listen {{ graphite_listen_port }}"
+    dest: '{{ httpd_config_parts_dir }}/15-graphite-web.conf'
+    create: yes
+    state: present
+    insertafter: "# Graphite Web Basic mod_wsgi vhost"
+    line: "Listen {{ graphite_listen_port }}"
```
Comment (larsks) [inline roles/grafana/tasks/graphite.yml:41]: You may want to make this conditional...if someone has specified `graphite_listen_port == 80`, then the `Listen` statement here would conflict with the default `Listen` statement.

Comment (larsks) [inline roles/grafana/tasks/graphite.yml]: Also, can the apache configuration here be made to work with the `opstools_vhost` role, which is how we're handling both uchiwa and kibana right now?  Maybe that is for a different PR, since this is mostly about firewall configuration...

**Review by larsks — CHANGES_REQUESTED** [review]
See inline comments

File: roles/firewall/commit/tasks/main.yml (hunk 7 above)
Comment (larsks) [reply roles/firewall/commit/tasks/main.yml]: Also, thinking about #79, we could make life easier for people by adding the rules to our own chain (say, `opstools_rules`), and then adding a single `-j opstools_rules` to the `INPUT` chain.  This will make it substantially easier to re-order firewall rules.  We may also want to consider inserting that rule at the _top_ of the input chain, rather than at the bottom, which may resolve the problem with our rules getting ignored because of a `DROP` or `REJECT` rule.

File: roles/elasticsearch/tasks/main.yml [hunk 12]
```diff
@@ -38,5 +38,13 @@
 
 - name: register elasticsearch firewall ports
   set_fact:
-    firewall_ports: >
-      {{ firewall_ports + [elasticsearch_port] }}
+    firewall_data: >
+      {{ firewall_data }} + [
+        {
+          'source': "{{ item }}",
```
Comment (larsks) [inline roles/elasticsearch/tasks/main.yml:44]: Shouldn't this be "destination"? The `elasticsearch_interface_ipv4` address is not where connections are coming _from_; it's where clients are connecting _to_.

File: roles/firewall/commit/tasks/main.yml [hunk 13]
```diff
@@ -1,20 +1,24 @@
...
 - name: Enable service ports via iptables
   iptables:
-      chain: INPUT
-      protocol: tcp
-      destination_port: "{{ item }}"
+      ip_version: "{% if item.source|is_ivp6 %}ipv6{% else %}ipv4{% endif %}"
+      chain: "{{ iptables_chain_name }}"
+      protocol: "{{ item.protocol }}"
+      destination_port: "{{ item.port }}"
+      source: "{{ item.source }}"
```
Comment (larsks) [inline roles/firewall/commit/tasks/main.yml:8]: `destination`

File: roles/firewall/commit/tasks/main.yml [hunk 14]
```diff
@@ -1,20 +1,24 @@
...
+  with_items: '{{ firewall_data }}'
   when: firewall_manage_rules and firewall_use_iptables
   notify: persist iptables
 
+
 - name: Enable service ports via firewalld
   firewalld:
-      port: "{{ item }}/tcp"
+      port: "{{ item.port }}/{{ item.protocol }}"
+      source: "{{ item.source }}"
```
Comment (larsks) [inline roles/firewall/commit/tasks/main.yml:18]: `destination`

**Review by larsks — CHANGES_REQUESTED** [review]
I am pretty sure you are using `source` when you should be using `destination`.

**Comment by paramite** [comment]
Closing in favor of  #89. This PR needs to be split in several few PRs.
